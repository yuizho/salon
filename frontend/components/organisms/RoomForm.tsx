import API, { GraphQLResult } from '@aws-amplify/api';
import { useRouter } from 'next/router';
import { FC } from 'react';
import { useRecoilState, useSetRecoilState } from 'recoil';
import { NETWORK_ERROR } from '../../graphql/error-message';
import { openRoom } from '../../graphql/mutations';
import { OpenRoomMutation } from '../../graphql/schema';
import { appState } from '../../states/app';
import { myState } from '../../states/me';
import Button from '../atoms/Button';

type Props = {
  onClick: () => void;
};

const mutateOpenRoom = async () => {
  const result = (await API.graphql({
    query: openRoom,
  })) as GraphQLResult<OpenRoomMutation>;
  return result;
};

export const Component: FC<Props> = ({ onClick }) => (
  <div className="flex flex-col space-y-6 border rounded p-8 ">
    <p className="text-slate-600">
      プランニングポーカーを開始するには部屋を作成してください。作成した部屋は約30分間有効です。
    </p>
    <Button text="部屋を作成する" onClick={onClick} />
  </div>
);

const Container: FC = () => {
  const router = useRouter();
  const setApp = useSetRecoilState(appState);
  const [, setMe] = useRecoilState(myState);

  const createRoom = async () => {
    try {
      const result = await mutateOpenRoom();
      const roomId = result.data?.openRoom.room_id ?? '';
      const userId = result.data?.openRoom.user_id ?? '';
      setMe({ roomId, userId });

      router.push(`/rooms/${roomId}`);
    } catch (e) {
      setApp((app) => ({
        ...app,
        loading: false,
        errorMessage: NETWORK_ERROR,
      }));
    }
  };

  return (
    <Component
      onClick={() => {
        createRoom();
      }}
    />
  );
};

export default Container;
