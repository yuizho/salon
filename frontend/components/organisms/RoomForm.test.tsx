import { render } from '@testing-library/react';
import { RecoilRoot } from 'recoil';

import RoomForm from './RoomForm';

describe('RoomForm', () => {
  test('snapshot', () => {
    const { asFragment } = render(
      <RecoilRoot>
        <RoomForm />
      </RecoilRoot>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
