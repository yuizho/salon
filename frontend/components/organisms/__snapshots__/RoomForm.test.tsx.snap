// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomForm snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col space-y-6 border rounded p-8 "
  >
    <p
      class="text-slate-600"
    >
      プランニングポーカーを開始するには部屋を作成してください。作成した部屋は約30分間有効です。
    </p>
    <p
      class="text-slate-600"
    >
      部屋を作成したら、URLを共有してメンバーを招待しましょう！
    </p>
    <button
      class="
      py-2 px-4
      font-semibold
      border-slate-400
      text-slate-600
      border
      rounded
      shadow-md
      bg-white
      w-full
      sm:w-auto
      sm:text-sm
      hover:border-2
      active:shadow-none
      active:border
      active:border-slate-300
    "
      type="button"
    >
      部屋を作成する
    </button>
  </div>
</DocumentFragment>
`;
